### ISSUE TITLE
Error on restart with v1 production case on chrysalis -- corruption in mosart restart history filenames

### ISSUE THREAD

**Issue by ndkeen** [E3SM-Project/E3SM#4175]
With our v1 hires production runs, when I change the PE layout to get around the issue https://github.com/E3SM-Project/E3SM/issues/4174, the runs seem to work fine -- ie completes and files written.  But the case will fail when attempting to read from restart files. The restart files themselves seem OK, but the code is trying to look for a filename that has wacky characters at the end (which looks like memory corruption).

```
1158:  ERROR:
1158:  GETFIL: FAILED to get ./20210318.A_WCYCL1950S_CMIP6_HR.ne120_oRRS18v3_ICG.unc06
1158:  .pd.mosart.rh0.0136-12-01-00000.nc^W{^[3^R\354\3453^A2\3452\373f\2262\366\313\2373^[0:3 \242l3)\324!3^P^?\3063^V

casedir:
/lcrc/group/e3sm/ac.ndkeen/scratch/chrys/E3SM_simulations/20210318.A_WCYCL1950S_CMIP6_HR.ne120_oRRS18v3_ICG.unc06.pd
```

I tried the following, all with same issue:
1) simply retrying a few times
2) running another case longer and restarting (problem only seems to happen when reading from restarts)
3) trying with and without threading (32x2 and 64x1) 
4) try rebuilding with DEBUG

The PE layouts I've tried so far that show this problem are both simple stacked layouts (128 nodes with 4096 MPI's using 32x2 and 64 nodes with 4096 MPI's using 64x1).

I tracked down the issue to a call: ncd_io() in ROF. Placing debug write calls inside `ncd_io_char_var2_nf`, I found:

```
3649:  ndk ncd_io_char_var2_nf read varname=locfnh
3649:  ndk ncd_io_char_var2_nf read data=
3649:  ./20210318.A_WCYCL1950S_CMIP6_HR.ne120_oRRS18v3_ICG.unc06.n064x1.mosart.h0.0136
3649:  -12.nc>Y9 ÷                                     
```

I don't yet know if it will be easy to find a smaller test (smaller resolution) that reproduces the issue. However, once the restart file are written, the fail happens in about 5 minutes.

**Comment by jonbob** [comment]
@ndkeen - it seems to me we've seen this before. Could it be related to this: [PR #3463](https://github.com/E3SM-Project/E3SM/pull/3463)?

**Comment by ndkeen** [comment]
Thanks @jonbob!  Looks promising, trying now.

**Comment by ndkeen** [comment]
That did not fix the issue when I simply rebuild and try to restart from files. Is it the case that I need to rerun to generate new restart files?

**Comment by jonbob** [comment]
I think so -- it seems to me the corruption is in the filename itself that gets written to the pointer file?

**Comment by ndkeen** [comment]
OK makes sense. Retrying a run now.

**Comment by ndkeen** [comment]
I think this fix does correct the problem. I re-ran the v1 hires production case for a month with this fix, then tried to restart and it seems to be working.

Note I was unable to find a simpler test that would show this error. Changing the default PIO_VERSION to be 2, I ran the following tests (without this fix) and all passed:

```
 ERS_Ld3.f45_f45.ICLM45ED.chrysalis_intel.clm-fates
 ERS_Ld31.ne4_ne4.FC5AV1C-L.chrysalis_intel
 ERS.f19_g16.I1850CNECACNTBC.chrysalis_intel.clm-eca
 ERS_Ld31_P1x1.1x1_smallvilleIA.ICLM45CNCROP.chrysalis_intel
 ERS.ne11_oQU240.A_WCYCL1850.chrysalis_intel
ERS_P2048_Ld3.ne120_oRRS18v3_ICG.A_WCYCL2000_H01AS.chrysalis_intel
```

**Comment by ndkeen** [comment]
This was fixed via PR#4180 to maint-1.0
